In [ ]:
# Refrence Video:- https://www.youtube.com/watch?v=6WDFfaYtN6s&t=291s

In [1]:
import pandas as pd
import numpy as np

## Selection only selecting columns

In [24]:
data=pd.read_csv("https://raw.githubusercontent.com/krishnaik06/Feature-Engineering-Live-sessions/master/mercedes.csv"
                 ,usecols=['X1','X2','X3','X4','X5','X6'])
data.head()

X1  X2 X3 X4 X5 X6
0  v  at  a  d  u  j
1  t  av  e  d  y  l
2  w   n  c  d  x  j
3  t   n  f  d  x  l
4  v   n  f  d  h  d

## Prinintg unique Values from all columns

In [3]:
for col in data.columns:
    print(col,":- ",len(data[col].unique()))

X1 :-  27
X2 :-  44
X3 :-  7
X4 :-  4
X5 :-  29
X6 :-  12


##### Lets examine how many cols do we get after aplying one hot encoding

In [7]:
pd.get_dummies(data,drop_first=True).shape

(4209, 117)

#### Here 117 means you are adding eaxtra 112 columns to this 5 columns
#### Note that as no. of features increases it give rise to curse of Dimensionality

https://proceedings.mlr.press/v7/niculescu09/niculescu09.pdf In the winning solution of KDD cup 2009, Winning the KDD cup orange chalange with Ensemble selection the author limit one hot encoding to the most 10 frequent labels of the variable. This mean that they would make one binary variable for each of the 10 most frequent labels only
. This is equivalent to grouping all the other labels under one category, i.e in this case will be dropped. Thus the 10 new dummy variable indicate if one of the 10 kost frequent label is prsent(1) or not(0) for a particular observations

Note: this 10 can varies according to the sitautions, it can be 20,30 or anything accrding to situation

### Etracting top 10 frequent values in column X2

In [4]:
data["X1"].value_counts().sort_values(ascending=False).head(20)

aa    833
s     598
b     592
l     590
v     408
r     251
i     203
a     143
c     121
o      82
w      52
z      46
u      37
e      33
m      32
t      31
h      29
y      23
f      23
j      22
Name: X1, dtype: int64

### Extracting index of top 10 values

In [5]:
top10=[x for x in data["X1"].value_counts().sort_values(ascending=False).head(10).index]
top10

['aa', 's', 'b', 'l', 'v', 'r', 'i', 'a', 'c', 'o']

In [14]:
for label in top10:
    data["label"]=np.where(data["X1"]==label,1,0)

In [23]:
data[data["label"]==1].head(5)

X1  X2 X3 X4 X5 X6  label
56   o  as  c  d  j  l      1
103  o  ag  f  d  j  j      1
113  o  as  f  d  j  l      1
474  o  as  f  d  d  l      1
671  o  ai  f  d  c  g      1

#### But we want the data in one hot encoding way i.e value should have become columns

In [7]:
data.drop("label",axis=1,inplace=True)

In [8]:
data.head()

X1  X2 X3 X4 X5 X6
0  v  at  a  d  u  j
1  t  av  e  d  y  l
2  w   n  c  d  x  j
3  t   n  f  d  x  l
4  v   n  f  d  h  d

## Lets make 10 binary variables

In [25]:
for label in top10:
    data[label]=np.where(data["X1"]==label,1,0)

In [26]:
data[["X1"]+top10].head(10)

X1  aa  s  b  l  v  r  i  a  c  o
0  v   0  0  0  0  1  0  0  0  0  0
1  t   0  0  0  0  0  0  0  0  0  0
2  w   0  0  0  0  0  0  0  0  0  0
3  t   0  0  0  0  0  0  0  0  0  0
4  v   0  0  0  0  1  0  0  0  0  0
5  b   0  0  1  0  0  0  0  0  0  0
6  r   0  0  0  0  0  1  0  0  0  0
7  l   0  0  0  1  0  0  0  0  0  0
8  s   0  1  0  0  0  0  0  0  0  0
9  b   0  0  1  0  0  0  0  0  0  0

#### Note here e is in top 10 i.e why its getting value 0 "at", "av" are not in top 10 so its 0

In [30]:
# Get whole set of data on dummy variables
def one_hot_encode(df,var,top_x_labels):
    # Function to create the dummy variables for most frequent variables
    # we can vary the number of most frequent labels that we can encode
    
    for labels in top_x_labels:
        df[var+"_"+labels]=np.where(df[var]==labels,1,0)

### For X2

In [31]:
# Read the data again
data=pd.read_csv("https://raw.githubusercontent.com/krishnaik06/Feature-Engineering-Live-sessions/master/mercedes.csv"
                 ,usecols=['X1','X2','X3','X4','X5','X6'])
top10=[x for x in data["X2"].value_counts().sort_values(ascending=False).head(10).index]

one_hot_encode(data,"X2",top10)
data.head()

X1  X2 X3 X4 X5 X6  X2_as  X2_ae  X2_ai  X2_m  X2_ak  X2_r  X2_n  X2_s  \
0  v  at  a  d  u  j      0      0      0     0      0     0     0     0   
1  t  av  e  d  y  l      0      0      0     0      0     0     0     0   
2  w   n  c  d  x  j      0      0      0     0      0     0     1     0   
3  t   n  f  d  x  l      0      0      0     0      0     0     1     0   
4  v   n  f  d  h  d      0      0      0     0      0     0     1     0   

   X2_f  X2_e  
0     0     0  
1     0     0  
2     0     0  
3     0     0  
4     0     0

### For X1

In [32]:

top10=[x for x in data["X1"].value_counts().sort_values(ascending=False).head(10).index]

one_hot_encode(data,"X1",top10)
data.head()

X1  X2 X3 X4 X5 X6  X2_as  X2_ae  X2_ai  X2_m  ...  X1_aa  X1_s  X1_b  X1_l  \
0  v  at  a  d  u  j      0      0      0     0  ...      0     0     0     0   
1  t  av  e  d  y  l      0      0      0     0  ...      0     0     0     0   
2  w   n  c  d  x  j      0      0      0     0  ...      0     0     0     0   
3  t   n  f  d  x  l      0      0      0     0  ...      0     0     0     0   
4  v   n  f  d  h  d      0      0      0     0  ...      0     0     0     0   

   X1_v  X1_r  X1_i  X1_a  X1_c  X1_o  
0     1     0     0     0     0     0  
1     0     0     0     0     0     0  
2     0     0     0     0     0     0  
3     0     0     0     0     0     0  
4     1     0     0     0     0     0  

[5 rows x 26 columns]

In [33]:
# For X3
top10=[x for x in data["X3"].value_counts().sort_values(ascending=False).head(10).index]

one_hot_encode(data,"X3",top10)
data.head()

X1  X2 X3 X4 X5 X6  X2_as  X2_ae  X2_ai  X2_m  ...  X1_a  X1_c  X1_o  X3_c  \
0  v  at  a  d  u  j      0      0      0     0  ...     0     0     0     0   
1  t  av  e  d  y  l      0      0      0     0  ...     0     0     0     0   
2  w   n  c  d  x  j      0      0      0     0  ...     0     0     0     1   
3  t   n  f  d  x  l      0      0      0     0  ...     0     0     0     0   
4  v   n  f  d  h  d      0      0      0     0  ...     0     0     0     0   

   X3_f  X3_a  X3_d  X3_g  X3_e  X3_b  
0     0     1     0     0     0     0  
1     0     0     0     0     1     0  
2     0     0     0     0     0     0  
3     1     0     0     0     0     0  
4     1     0     0     0     0     0  

[5 rows x 33 columns]

In [34]:
# For X4
top10=[x for x in data["X4"].value_counts().sort_values(ascending=False).head(10).index]

one_hot_encode(data,"X4",top10)
data.head()

X1  X2 X3 X4 X5 X6  X2_as  X2_ae  X2_ai  X2_m  ...  X3_f  X3_a  X3_d  X3_g  \
0  v  at  a  d  u  j      0      0      0     0  ...     0     1     0     0   
1  t  av  e  d  y  l      0      0      0     0  ...     0     0     0     0   
2  w   n  c  d  x  j      0      0      0     0  ...     0     0     0     0   
3  t   n  f  d  x  l      0      0      0     0  ...     1     0     0     0   
4  v   n  f  d  h  d      0      0      0     0  ...     1     0     0     0   

   X3_e  X3_b  X4_d  X4_a  X4_b  X4_c  
0     0     0     1     0     0     0  
1     1     0     1     0     0     0  
2     0     0     1     0     0     0  
3     0     0     1     0     0     0  
4     0     0     1     0     0     0  

[5 rows x 37 columns]

In [35]:
# For X5
top10=[x for x in data["X5"].value_counts().sort_values(ascending=False).head(10).index]

one_hot_encode(data,"X5",top10)
data.head()

X1  X2 X3 X4 X5 X6  X2_as  X2_ae  X2_ai  X2_m  ...  X5_w  X5_v  X5_q  X5_r  \
0  v  at  a  d  u  j      0      0      0     0  ...     0     0     0     0   
1  t  av  e  d  y  l      0      0      0     0  ...     0     0     0     0   
2  w   n  c  d  x  j      0      0      0     0  ...     0     0     0     0   
3  t   n  f  d  x  l      0      0      0     0  ...     0     0     0     0   
4  v   n  f  d  h  d      0      0      0     0  ...     0     0     0     0   

   X5_d  X5_s  X5_n  X5_p  X5_m  X5_i  
0     0     0     0     0     0     0  
1     0     0     0     0     0     0  
2     0     0     0     0     0     0  
3     0     0     0     0     0     0  
4     0     0     0     0     0     0  

[5 rows x 47 columns]

In [36]:
# For X6
top10=[x for x in data["X6"].value_counts().sort_values(ascending=False).head(10).index]

one_hot_encode(data,"X6",top10)
data.head()

X1  X2 X3 X4 X5 X6  X2_as  X2_ae  X2_ai  X2_m  ...  X6_g  X6_j  X6_d  X6_i  \
0  v  at  a  d  u  j      0      0      0     0  ...     0     1     0     0   
1  t  av  e  d  y  l      0      0      0     0  ...     0     0     0     0   
2  w   n  c  d  x  j      0      0      0     0  ...     0     1     0     0   
3  t   n  f  d  x  l      0      0      0     0  ...     0     0     0     0   
4  v   n  f  d  h  d      0      0      0     0  ...     0     0     1     0   

   X6_l  X6_a  X6_h  X6_k  X6_c  X6_b  
0     0     0     0     0     0     0  
1     1     0     0     0     0     0  
2     0     0     0     0     0     0  
3     1     0     0     0     0     0  
4     0     0     0     0     0     0  

[5 rows x 57 columns]

In [39]:
#Now we dont need var x1 to X6 drop these
data.drop(["X1","X2","X3","X4","X5","X6"],axis=1,inplace=True)

In [40]:
data.shape

(4209, 51)

#### Now see we reduce the features from 117 to 51